File: Data Structures & Algorithms/max-area-of-island/submission-0.cs
public class Solution {
    public int MaxAreaOfIsland(int[][] grid) {
        if (grid == null) return 0;
        int result = 0;

        for (int i = 0; i < grid.Length; i++) {
            for (int j = 0; j < grid[0].Length; j++) {
                if (grid[i][j] == 1) {
                    result = Math.Max(dfs(grid, i, j), result);
                }
            }
        }
        return result;
    }

    private int dfs(int[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.Length || j >= grid[0].Length || grid[i][j] != 1) {
            return 0;
        }

        grid[i][j] = 0;

        return 1 + dfs(grid, i - 1, j) + dfs(grid, i + 1, j) + dfs(grid, i, j - 1) + dfs(grid, i, j + 1);
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {

        int[] prefixArray = calculatePrefixArray(height);
        int[] suffixArray = calculateSuffixArray(height);
        int result = 0;
        for (int i = 0; i < height.length; i++){
            int waterAmount = Math.min(prefixArray[i], suffixArray[i]) - height[i];
            result += waterAmount > 0 ? waterAmount : 0;
        }
        return result;
    }

    private static int[] calculatePrefixArray(int[] height){
        int maxLeft = height[0];
        int[] arr = new int[height.length];
        for(int i=0; i < height.length; i++ ){
            maxLeft = Math.max(maxLeft, height[i]);
            arr[i] = maxLeft;
        }
        return arr;
    }

    private static int[] calculateSuffixArray(int[] height){
        int maxLeft = height[height.length - 1];
        int[] arr = new int[height.length];
        for(int i=height.length -1; i > 0; i-- ){
            maxLeft = Math.max(maxLeft, height[i]);
            arr[i] = maxLeft;
        }
        return arr;
    }

    
}